program/rust/src/processor/upd_price.rs

```diff
@@ -127,27 +127,74 @@ pub fn upd_price(
     // Check clock
     let clock = Clock::from_account_info(clock_account)?;
 
-    let mut publisher_index: usize = 0;
+    let publisher_index: usize;
     let latest_aggregate_price: PriceInfo;
 
     // The price_data borrow happens in a scope because it must be
     // dropped before we borrow again as raw data pointer for the C
     // aggregation logic.
     {
         // Verify that symbol account is initialized
-        let price_data = load_checked::<PriceAccount>(price_account, cmd_args.header.version)?;
+        let mut price_data = load_checked::<PriceAccount>(price_account, cmd_args.header.version)?;
 
-        // Verify that publisher is authorized
-        while publisher_index < try_convert::<u32, usize>(price_data.num_)? {
-            if price_data.comp_[publisher_index].pub_ == *funding_account.key {
-                break;
+        // Verify that publisher is authorized by initially binary searching
+        // for the publisher's component in the price account. The binary
+        // search might not work if the publisher list is not sorted; therefore
+        // we fall back to a linear search and during the search we try to bring
+        // the publishers list closer to a sorted state. (We do not sort them
+        // entirely to avoid exceeding the compute budget.)
+        let mut binary_search_result = None;
+
+        {
+            // Binary search to find the publisher key. Rust std binary search is not used because
+            // they guarantee valid outcome only if the array is sorted whereas we want to rely on
+            // a Equal match if it is a result on an unsorted array. Currently the rust
```

Comment:
Is this still true if even if you use the binary_search_by_key variants? My worry in the call was the the result can't be verified but you mentioned the quality should confirm. Just want to double check if the reason the removal those forms of binary_search* is for this reason? The quality match branch seems right.

Reply:
Well the implementation is the same but the API doc doesn't guarantee that. That's why I want to just be on the safe side here by having our code.

Reply:
To be specifc the rust interface doesn't guarantee that if it returns an Ok(index) the pub[index] == key and also it holds true based on their implementation, i don't want to rely on it (who knows, maybe they change it later and we don't notice it).